keps/sig-network/2999-host-network-pod-ips/README.md

```diff
@@ -0,0 +1,603 @@
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Host-network pods inherit their Pod IPs from the node where they are running. Nodes may change their IPs during their lifecycle, and kubelet updates the node.Status accordingly, however, host-network pods IPs are not updated and keep reporting the old IPs until they are restarted/recreated.
```

Comment:
> Host-network pods inherit their Pod IPs from the node where they are running.

I have never found good documentation on this, but there are some considerations for a host with multiple NICs with regard to `hostNet: true` pods. Most Linux network managers (tested: `systemdnetworkd` and `NetworkManager`) enable you to define a default route. So far I have only seen pods pull the same IP based on the default route. Although I can't back this up with paperwork. I haven't looked that far into the code (yet?).

However I am wondering how this change would respond in the following situations given this observation. 

Given a node with the following criteria 

 - eno1 `10.0.0.100` (default route)
 - eno2 `192.168.1.1
 

 - What is the expected Pod behavior if the default route changes from `eno1` -> `eno2`?
 - What is the expected Pod behavior if `eno1` is disabled (deactivated) and `eno2` has a valid connection, but is not the default route?
  - What is the expected Pod behavior if a new NIC is added, but the default route remains?
  - What is the expected Pod behavior if `eno1` at `10.0.0.100` remains the default route in the route table, but is assigned a new IP address `172.0.100.100`?
  
  tldr; that default route though

Reply:
There are two types of node ip changes, in context:
1. **Soft change** This happens because the delta between what kubelet sees (default first IPv4 iirc) and thus Static Pods see (if they start before cloud provider set-ip step) and what a thing like cloud provider sets which can be dual-stacked, different order, or even different IP.
2.  **Hard change** Like the one you described or what @aojea described (e.g., DHCP resets) or to be more specific a reconfiguration on the networking stack of the node.

The first type has no baring on anything. It just says the pod is reachable using different set of addresses (assuming pod listens to `::0` or whatever equivalent in other languages). The node itself didn't change its own network stack configuration. To be more specific. The node **only** changed the IPs it reports to kubernetes from `{<IP>}` to `!{<IP>}` but the node owned and still owns the joined set of IPs. Please do note that srcIP for host-net pod does have to == pod IP, this is subject to how node routing is configured (similar to pod IPs this is subject CNI and its snat rules). This is the most common change we see.

For the second type also - arguably - should result into no change in pod (again assuming it listens to `::0` and it shares host network namespace). To present it differently let us assume the node retooled the entire networking stack (assuming CNI is smart enough to deal with this and no problems with pods in pod-net) what happens to kubelet? in practice kubelet is running as a pod in pod-net (or as any other app in host default net namespace). Depending on the severity of change existing connection may need to be terminated and re-established but it does not require kubelet to restart (again assuming listening to `::0`).

Maybe we should limit the kep to first type? @aojea

Reply:
> Although I can't back this up with paperwork. I haven't looked that far into the code (yet?).

the code is here if you want to have some fun :) https://github.com/kubernetes/kubernetes/blob/6ee6251fea608ff64ff69314624d26fc6b3b61ac/pkg/kubelet/nodestatus/setters.go#L60-L255

There are different scenarios, let's discard the cloud provider one and focus on bare metal/non cloud-provider, the code says
```

 			// 1) Use nodeIP if set (and not "0.0.0.0"/"::")
			// 2) If the user has specified an IP to HostnameOverride, use it
			// 3) Lookup the IP from node name by DNS
			// 4) Try to get the IP from the network interface used as default gateway
			//
			// For steps 3 and 4, IPv4 addresses are preferred to IPv6 
```
if we are in step 4, we execute https://github.com/kubernetes/kubernetes/blob/6ee6251fea608ff64ff69314624d26fc6b3b61ac/staging/src/k8s.io/apimachinery/pkg/util/net/interface.go#L463-L468
```
// ResolveBindAddress returns the IP address of a daemon, based on the given bindAddress:
// If bindAddress is unset, it returns the host's default IP, as with ChooseHostInterface().
// If bindAddress is unspecified or loopback, it returns the default IP of the same
// address family as bindAddress.
// Otherwise, it just returns bindAddress.
func ResolveBindAddress(bindAddress net.IP) (net.IP, error) {
```

It is important to mention that these are host-network pods, so they have ALL the IPs in the host, they run in the host namespace, we are talking only about the pod.Status.PodIPs fields, there is no influence in the node network at all.

@khenidak I narrowed the scope of the KEP in goals/non-goals to avoid any change in the kubelet. This is just to set the reported nodeIP by the kubelet  in the PodIPs, 

You can see Dan's comments too, the whole IP dance inside of the kubelet requires a big big big work if we want to make it perfect
https://github.com/kubernetes/enhancements/pull/3000#discussion_r721361112
https://github.com/kubernetes/enhancements/pull/3000#discussion_r721369269